### PR TITLE
Prevent External Users Registration

### LINKED ISSUE #17048: External Users are being able to register a new account even when DisableNewRegistrations is enabled
### Describe the bug

When using external authentication feature, users login from external providers still create a new account even when the setting `DisableNewRegistrations` is set to true.

### Orchard Core version

v2.1

### To Reproduce

1. Enable any external authentication provider feature.
2. Set DisableNewRegistrations to true in the External Authentication settings.
3. Try to login using the external provider.

### Expected behavior
Users should not be able to register if `DisableNewRegistrations` is set to true.

### LINKED ISSUE #17059: External Authentication Feature is causing an infinite redirect.
### Describe the bug
When a user login from external provider, but creating the account fails we redirect them back to the login page. But if the "Use external provider for login" settings is set, the user will be caught in an infinite loop redirecting back and fourth.

### Orchard Core version

2.1

### To Reproduce

<!-- Steps to reproduce the behavior:
1. Enable external provider feature
2. In Security >> User Registration >> check "Disable new external registration"
3. In Security >> User Login >> Click on the External Login tab. Enable "Use external provider for login" options.
4. Try to login with a new user that does not exists on the local site using the external provider.

### Expected behavior

We should see an error not causing an infinite loop.

### PR DESCRIPTION
**Pull request by MikeAlhayek**
### Summary

Fix #17048
Fix #17059

This PR also refines the registration and login processes to enable event-based hooks, allowing for easier customization and decoupling of code based on the enabled features. It also improves feature dependencies so code like the following is no longer needed:

```
var registrationFeatureIsAvailable = (await _shellFeaturesManager.GetAvailableFeaturesAsync())
                       .Any(feature => feature.Id == UserConstants.Features.UserRegistration);

// do something...
```

### Release Notes
The user registration and login functionalities have been refactored for better extensibility:

- **Registration Improvements**:  
  - The `IRegistrationFormEvents` interface now includes `Task RegisteringAsync(UserRegisteringContext context)` for streamlined customization.  
  - A new base class, `RegistrationFormEventsBase`, allows developers to override only necessary methods.

- **Login Improvements**:  
  - The `ILoginFormEvent` interface has a new method: `Task<IActionResult> ValidatingLoginAsync(IUser user)`.  
  - The `LoginFormEventBase` class enables overriding relevant methods. Note that the base implementation of `LoggingInAsync()` has been removed; you must now implement this method if using `LoginFormEventBase`.

- **User Service Update**:  
  - A new method in `IUserService` interface: `Task<IUser> RegisterAsync(RegisterUserForm model, Action<string, string> reportError)` facilitates registration with error reporting.

These enhancements make the user management system more modular and customizable.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Users.Core/Models/LoginSettings.cs [hunk 1]
```diff
@@ -4,6 +4,8 @@ namespace OrchardCore.Users.Models;
 
 public class LoginSettings
 {
+    public bool UsersMustValidateEmail { get; set; }
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Users.Core/Models/LoginSettings.cs:7]: Why is this better here rather than `RegistrationOptions`? Email validation happens during registration. Even if that registration happens during an external login, it's still a registration.

File: test/OrchardCore.Tests/OrchardCore.Tests.csproj [hunk 2]
```diff
@@ -87,6 +85,10 @@
     </None>
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="OrchardCore.Users\Recipes\" />
+  </ItemGroup>
```
Comment (Piedone) [inline test/OrchardCore.Tests/OrchardCore.Tests.csproj:90]: Why is this needed?

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationUserEvent.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.AspNetCore.Routing;
+using OrchardCore.Mvc.Core.Utilities;
+using OrchardCore.Settings;
+using OrchardCore.Users.Controllers;
+using OrchardCore.Users.Events;
+using OrchardCore.Users.Models;
+
+namespace OrchardCore.Users.Handlers;
+
+internal sealed class EmailConfirmationUserEvent : LoginFormEventBase
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationUserEvent.cs:13]:
```suggestion
internal sealed class EmailConfirmationLoginFormEvent : LoginFormEventBase
```

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/RegistrationUserEvent.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.AspNetCore.Routing;
+using OrchardCore.Mvc.Core.Utilities;
+using OrchardCore.Settings;
+using OrchardCore.Users.Controllers;
+using OrchardCore.Users.Events;
+using OrchardCore.Users.Models;
+
+namespace OrchardCore.Users.Handlers;
+
+internal sealed class RegistrationUserEvent : LoginFormEventBase
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Handlers/RegistrationUserEvent.cs:13]:
```suggestion
internal sealed class UserModerationLoginFormEvent : LoginFormEventBase
```

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/UserRegistrationFormEvents.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+using OrchardCore.Settings;
+using OrchardCore.Users.Events;
+using OrchardCore.Users.Models;
+
+namespace OrchardCore.Users.Handlers;
+
+internal sealed class UserRegistrationFormEvents : RegistrationFormEventsBase
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Handlers/UserRegistrationFormEvents.cs:7]:
```suggestion
internal sealed class UserModerationRegistrationFormEvents : RegistrationFormEventsBase
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationRegistrationFormEvents.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    private readonly HtmlEncoder _htmlEncoder;
+
+    internal readonly IStringLocalizer S;
+
+    public EmailConfirmationRegistrationFormEvents(
+        ISiteService siteService,
+        UserManager<IUser> userManager,
+        LinkGenerator linkGenerator,
+        IHttpContextAccessor httpContextAccessor,
+        IEmailService emailService,
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationRegistrationFormEvents.cs:34]: Good point. Maybe we need to register this handler only when the Email feature is enabled. In fact anything that is related to email confirmation should depend on the Email feature.

File: src/OrchardCore/OrchardCore.Users.Core/Models/LoginSettings.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Models/LoginSettings.cs:7]: The problem is that RegistrationSettings should only be accessed when the registration feature. We should never try to access it from the main feature. Moving it to the LoginSettings is better because we can access it from the main feature. Also, this setting is check during registration and login so it is not only for registration but for the core feature of the users feature.

File: test/OrchardCore.Tests/OrchardCore.Tests.csproj (hunk 2 above)
Comment (MikeAlhayek) [reply test/OrchardCore.Tests/OrchardCore.Tests.csproj:90]: It's not.  I think because I removed the recipe file this was added because we now have empty folder. I'll remove it

File: src/OrchardCore/OrchardCore.Users.Abstractions/Events/ILoginFormEvent.cs [hunk 7]
```diff
@@ -42,6 +42,11 @@ public interface ILoginFormEvent
...
     /// <returns></returns>
-    Task<IActionResult> LoggingInAsync()
-        => Task.FromResult<IActionResult>(null);
+    Task<IActionResult> LoggingInAsync();
+
+    /// <summary>
+    /// Occurs when is trying to login and after the user is located.
+    /// </summary>
+    /// <returns></returns>
+    Task<IActionResult> LoggingInAsync(IUser user);
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Abstractions/Events/ILoginFormEvent.cs:51]: I am okay with changing the name. But I am not fan of the suggested name. Better alternative?

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationRegistrationFormEvents.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationRegistrationFormEvents.cs:34]: fixed.

File: test/OrchardCore.Tests/OrchardCore.Tests.csproj (hunk 2 above)
Comment (MikeAlhayek) [reply test/OrchardCore.Tests/OrchardCore.Tests.csproj]: removed

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/RegistrationUserEvent.cs (hunk 4 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Handlers/RegistrationUserEvent.cs]: done

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/UserRegistrationFormEvents.cs (hunk 5 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Handlers/UserRegistrationFormEvents.cs]: done

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationUserEvent.cs (hunk 3 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Handlers/EmailConfirmationUserEvent.cs]: done

File: src/OrchardCore/OrchardCore.Users.Abstractions/Events/ILoginFormEvent.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Abstractions/Events/ILoginFormEvent.cs]: I renamed it to `ValidatingLoginAsync`

File: src/OrchardCore/OrchardCore.Users.Core/Models/LoginSettings.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Models/LoginSettings.cs]: I made other changes to ensure and undone this change.

**Comment by Piedone** [comment]
Please don't resolve the convos, because it gets really hard for me to follow them for a second review.

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 8]
```diff
@@ -127,37 +127,45 @@ public async Task<IActionResult> LoginPOST(string returnUrl = null)
...
+                        var loginResult = await accountEvent.ValidatingLoginAsync(user);
 
                         if (loginResult != null)
                         {
                             return loginResult;
                         }
                     }
 
-                    result = await _signInManager.PasswordSignInAsync(user, model.Password, model.RememberMe, lockoutOnFailure: true);
+                    await _accountEvents.InvokeAsync((e, user, modelState) => e.LoggingInAsync(user.UserName, (key, message) => modelState.AddModelError(key, message)), user, ModelState, _logger);
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:143]: This is now run twice, also up on ln 119. Is this deliberate?

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 8 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:143]: It's not. I'll fix that tomorrow.

File: src/OrchardCore.Modules/OrchardCore.Users/Services/ExternalLoginFormEvents.cs [hunk 9]
```diff
@@ -36,10 +42,30 @@ public override async Task<IActionResult> LoggingInAsync()
...
 
+        // To prevent infinite redirects, we store an ELAR cookie, which stands for ExternalLoginAutoRedirect.
+        // This cookie helps us avoid executing the same redirect multiple times. 
+        // The acronym ELAR has no special meaning, and any other value could be used in its place; it is chosen simply for brevity.
+        if (_httpContextAccessor.HttpContext.Request.Cookies.TryGetValue(ExternalLoginAutoRedirectCookieName, out var _))
+        {
+            _httpContextAccessor.HttpContext.Response.Cookies.Delete(ExternalLoginAutoRedirectCookieName);
+
+            return null;
+        }
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Users/Services/ExternalLoginFormEvents.cs:53]: When can such an infinite redirect happen, why is this cookie necessary? I tried external login with Entra ID before this change and it didn't cause a redirect loop (though `UseExternalProviderIfOnlyOneDefined` was on the default `false`, i.e. this method short-circuited).

Also, instead of cookies, you can add a query string parameter, which might be safer than an external login (cookies might not be accepted), and we're not saving something in a cookie that needn't be persistent beyond the next request.

Additionally, if one signs out and then in again in the same browser (or just gets signed out when closing it, given that this cookie will persist beyond browser close but the user session doesn't) the cookie will remain there, and the logic form line 69 won't run, but I think it should. I think this will be the case too if a login provider is added and the user wants to use it but won't be able to.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Services/ExternalLoginFormEvents.cs (hunk 9 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Services/ExternalLoginFormEvents.cs:53]: The issue is explained in #17059. I can change this into using query params.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Users/Services/ExternalLoginFormEvents.cs:53]: I see, OK then. Yes, I'd prefer a query string parameter (or perhaps even `TempData`, but I'm wary of that) over cookies.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Services/ExternalLoginFormEvents.cs]: @Piedone query-parameter did not work, It gets lost in the redirects. I did use TempData instead. Let me know if you see anything else or if you approve this.

**Review by Piedone — APPROVED** [review]
